File: js/src/toolsEdit.ts
/**
 * Edit-surface tools — docx_replace / docx_edit_paragraph / docx_insert /
 * docx_delete / docx_revision — with result shapes per spec/tools/*.json.
 *
 * Every edit validates its anchor hash FIRST (algorithms.md §1): hash mismatch
 * → anchor_stale, ordinal out of range → anchor_not_found, unparseable →
 * anchor_invalid. Errors are thrown as ToolError (spec/errors.json codes).
 * Behavior mirrors `python/src/docxengine/_tools_edit.py` exactly (§6a).
 */
import type { AnchorEntry } from "./anchors.js";
import {
  type Revision,
  type RevisionFilter,
  applyPlainMatch,
  applyTrackedReplace,
  diffBlocks,
  diffUnits,
  findOccurrences,
  maxRevisionId,
  mergeParagraphRuns,
  paragraphBlock,
  rebuildParagraph,
  resolveRevisions,
  revisionAuthor,
  revisionDate,
  revisionMatches,
  revisionOpen,
  scanRevisions,
  tToDelText,
  wordDiff,
} from "./edits.js";
import { ToolError } from "./errors.js";
import { parseStyles } from "./projector.js";
import type { DocHandle, Session } from "./session.js";
import {
  type SpliceEdit,
  WS_RUN_RE,
  childElements,
  emitTextElement,
  escapeAttr,
  splice,
  spliceAll,
  textWithMap,
} from "./xmlscan.js";

type ResponseFormat = "concise" | "detailed";

// ---------------------------------------------------------------------------
// Shared helpers (anchor grammar + §6a validation order: parse → range → hash)
// ---------------------------------------------------------------------------

/** Full edit-grade anchor: edits always validate the hash (§6a). */
const FULL_ANCHOR_RE = /^P([1-9][0-9]*)#([0-9a-f]{4})$/;
/** Paragraph range; endpoint hashes are validated when present (§6a). */
const RANGE_RE = /^P([1-9][0-9]*)(?:#([0-9a-f]{4}))?\.\.P([1-9][0-9]*)(?:#([0-9a-f]{4}))?$/;

function plural(n: number, noun: string): string {
  return n === 1 ? `${n} ${noun}` : `${n} ${noun}s`;
}

function anchorInvalidError(detail: string): ToolError {
  return new ToolError("anchor_invalid", detail, [
    "Check the format 'P{index}#{hash}' (ranges: 'P10..P24').",
  ]);
}

function anchorNotFoundError(label: string): ToolError {
  return new ToolError("anchor_not_found", `Anchor ${label} not found: index out of range.`, [
    "Call docx_outline to re-map anchors.",
  ]);
}

function anchorStaleError(anchor: string): ToolError {
  return new ToolError(
    "anchor_stale",
    `Anchor ${anchor} is stale: the hash no longer matches the paragraph content.`,
    ["Call docx_read {anchor, window} and retry with the fresh anchor."],
  );
}

/** Body paragraphs only, indexable by `ordinal - 1`. */
function paragraphEntries(doc: DocHandle): AnchorEntry[] {
  return doc.anchorIndex().filter((e) => e.kind === "p");
}

function entryAt(entries: readonly AnchorEntry[], ordinal: number, label: string): AnchorEntry {
  const entry = entries[ordinal - 1];
  if (entry === undefined) throw anchorNotFoundError(label);
  return entry;
}

/** §6a validation order: parse → ordinal in range → hash match. */
function requireParagraph(entries: readonly AnchorEntry[], anchor: string): AnchorEntry {
  const m = FULL_ANCHOR_RE.exec(anchor);
  if (!m) throw anchorInvalidError(`Malformed anchor string: ${anchor}.`);
  const entry = entryAt(entries, Number(m[1]), anchor);
  if (entry.anchor !== anchor) throw anchorStaleError(anchor);
  return entry;
}

function commit(doc: DocHandle, xml: string): void {
  doc.pkg.setPart(doc.documentPartName, xml);
  doc.invalidate();
}

function freshParagraphAnchor(doc: DocHandle, ordinal: number): string {
  return entryAt(paragraphEntries(doc), ordinal, `P${ordinal}`).anchor;
}

// ---------------------------------------------------------------------------
// docx_replace (§4 coalescing, §5 tracking, §6a result shape)
// ---------------------------------------------------------------------------

export interface DocxReplaceArgs {
  doc_id: string;
  anchor?: string | undefined;
  old: string;
  new: string;
  all?: boolean | undefined;
  track_changes?: boolean | undefined;
  author?: string | undefined;
  response_format?: ResponseFormat | undefined;
}

export interface DocxReplaceResult {
  n_replaced: number;
  new_anchor?: string;
  anchors?: string[];
}

export function docxReplace(session: Session, args: DocxReplaceArgs): DocxReplaceResult {
  const doc = session.get(args.doc_id);
  let xml = doc.documentXml();
  const entries = paragraphEntries(doc);
  // Anchor hash validated FIRST (§1) — before any matching or argument checks.
  const target = args.anchor != null ? requireParagraph(entries, args.anchor) : null;
  if (!args.old) {
    throw new ToolError("not_found", "Text not found: the search text is empty.", [
      "Provide non-empty old text.",
    ]);
  }
  const all = args.all === true;
  const tracked = args.track_changes === true;
  const ordinals = target !== null ? [target.ordinal] : entries.map((e) => e.ordinal);

  // §6a: matches are non-overlapping, left-to-right, found per paragraph
  // before any splice. §5 ids are allocated in emission (document) order.
  interface Job {
    ordinal: number;
    s: number;
    e: number;
    delId: number;
    insId: number | null;
  }
  const jobs: Job[] = [];
  let next = maxRevisionId(xml);
  for (const ordinal of ordinals) {
    const entry = entries[ordinal - 1] as AnchorEntry;
    const { text } = textWithMap(xml, entry.block);
    for (const [s, e] of findOccurrences(text, args.old)) {
      jobs.push({
        ordinal,
        s,
        e,
        delId: tracked ? ++next : 0,
        insId: tracked && args.new !== "" ? ++next : null,
      });
    }
  }
  if (!all) {
    if (jobs.length === 0) {
      throw new ToolError("not_found", `Text not found: ${args.old}.`, [
        "Broaden the query; check the projection for the exact text.",
      ]);
    }
    if (jobs.length > 1) {
      throw new ToolError(
        "ambiguous_target",
        `${args.old} matches ${jobs.length} times without all: true.`,
        ["Add all: true or narrow with an anchor."],
      );
    }
  }
  if (jobs.length === 0) return { n_replaced: 0, anchors: [] }; // idempotent (§6a)

  const author = revisionAuthor(args.author);
  const date = revisionDate();
  // Apply matches in reverse document order so earlier offsets stay valid.
  for (const j of [...jobs].reverse()) {
    const block = paragraphBlock(xml, j.ordinal);
    xml = tracked
      ? applyTrackedReplace(xml, block, j.s, j.e, args.new, author, date, {
          delId: j.delId,
          insId: j.insId,
        })
      : applyPlainMatch(xml, block, j.s, j.e, args.new);
  }
  commit(doc, xml);
  const affected = [...new Set(jobs.map((j) => j.ordinal))].sort((a, b) => a - b);
  if (all) {
    // §6a: all → anchors of affected paragraphs ascending; otherwise new_anchor.
    return { n_replaced: jobs.length, anchors: affected.map((o) => freshParagraphAnchor(doc, o)) };
  }
  const ordinal = target !== null ? target.ordinal : (affected[0] as number);
  return { n_replaced: jobs.length, new_anchor: freshParagraphAnchor(doc, ordinal) };
}

// ---------------------------------------------------------------------------
// docx_edit_paragraph (§6 LCS word diff, §6a)
// ---------------------------------------------------------------------------

export interface DocxEditParagraphArgs {
  doc_id: string;
  anchor: string;
  text: string;
  track_changes?: boolean | undefined;
  author?: string | undefined;
  response_format?: ResponseFormat | undefined;
}

export interface DocxEditParagraphResult {
  new_anchor: string;
  diff: string;
}

export function docxEditParagraph(
  session: Session,
  args: DocxEditParagraphArgs,
): DocxEditParagraphResult {
  const doc = session.get(args.doc_id);
  const xml = doc.documentXml();
  const entry = requireParagraph(paragraphEntries(doc), args.anchor); // hash validated FIRST
  const { text: raw } = textWithMap(xml, entry.block);
  const ops = wordDiff(diffUnits(raw), diffUnits(args.text));
  const blocks = diffBlocks(ops);
  const newXml = rebuildParagraph(xml, entry.block, blocks, {
    tracked: args.track_changes === true,
    author: revisionAuthor(args.author),
    date: revisionDate(),
  });
  commit(doc, newXml);
  // §6a: n = max(#deleted units, #inserted units); noun singular when n is 1.
  const changed = Math.max(
    ops.filter((o) => o.op === "del").length,
    ops.filter((o) => o.op === "ins").length,
  );
  return {
    new_anchor: freshParagraphAnchor(doc, entry.ordinal),
    diff: `~${plural(changed, "word")} changed`,
  };
}

// ---------------------------------------------------------------------------
// docx_insert (§6a minimal markdown)
// ---------------------------------------------------------------------------

export interface DocxInsertArgs {
  doc_id: string;
  after?: string | undefined;
  before?: string | undefined;
  content: string;
  style?: string | undefined;
  track_changes?: boolean | undefined;
  author?: string | undefined;
  response_format?: ResponseFormat | undefined;
}

export interface DocxInsertResult {
  new_anchors: string[];
}

interface MdBlock {
  text: string;
  style?: string | undefined;
}

const MD_HEADING_RE = /^(#{1,9}) /;

/** §6a minimal markdown: one paragraph per non-blank line. */
export function parseMinimalMarkdown(content: string): MdBlock[] {
  const out: MdBlock[] = [];
  for (const raw of content.split("\n")) {
    const line = raw.endsWith("\r") ? raw.slice(0, -1) : raw;
    if (line.replace(WS_RUN_RE, "") === "") continue;
    const h = MD_HEADING_RE.exec(line);
    if (h) {
      const level = (h[1] as string).length;
      out.push({ text: line.slice(level + 1), style: `Heading${level}` });
    } else if (line.startsWith("- ") || line.startsWith("* ")) {
      out.push({ text: line.slice(2), style: "ListParagraph" }); // numPr wiring is Phase 2
    } else {
      out.push({ text: line });
    }
  }
  return out;
}

/** §6a: the styleId verbatim if defined, else with whitespace removed, else error. */
function resolveStyleId(doc: DocHandle, style: string): string {
  const styles = parseStyles(
    doc.pkg.has("word/styles.xml") ? doc.pkg.partText("word/styles.xml") : undefined,
  );
  if (styles.has(style)) return style;
  const compact = style.replace(WS_RUN_RE, "");
  if (styles.has(compact)) return compact;
  throw new ToolError("style_unknown", `Named style ${style} does not exist.`, [
    'Call docx_style {op: "list"} to see available styles.',
  ]);
}

export function docxInsert(session: Session, args: DocxInsertArgs): DocxInsertResult {
  const doc = session.get(args.doc_id);
  if ((args.after == null) === (args.before == null)) {
    throw anchorInvalidError("Provide exactly one of after or before.");
  }
  const entries = paragraphEntries(doc);
  const entry = requireParagraph(entries, (args.after ?? args.before) as string); // hash FIRST
  const paragraphs = parseMinimalMarkdown(args.content);
  const styleId = args.style != null ? resolveStyleId(doc, args.style) : null;
  if (paragraphs.length === 0) return { new_anchors: [] };

  const xml = doc.documentXml();
  const tracked = args.track_changes === true;
  const author = revisionAuthor(args.author);
  const date = revisionDate();
  let revId = tracked ? maxRevisionId(xml) + 1 : 0;
  const pieces: string[] = [];
  for (const b of paragraphs) {
    const effective = styleId ?? b.style;
    const pPr =
      effective != null ? `<w:pPr><w:pStyle w:val="${escapeAttr(effective)}"/></w:pPr>` : "";
    let run = b.text !== "" ? `<w:r>${emitTextElement("w:t", b.text)}</w:r>` : "";
    if (tracked && run !== "") {
      run = revisionOpen("ins", revId++, author, date) + run + "</w:ins>";
    }
    pieces.push(`<w:p>${pPr}${run}</w:p>`);
  }
  const position = args.after != null ? entry.block.end : entry.block.start;
  commit(doc, splice(xml, position, position, pieces.join("")));
  const base = args.after != null ? entry.ordinal + 1 : entry.ordinal;
  const fresh = paragraphEntries(doc);
  return {
    new_anchors: paragraphs.map((_, i) => entryAt(fresh, base + i, `P${base + i}`).anchor),
  };
}

// ---------------------------------------------------------------------------
// docx_delete (§6a)
// ---------------------------------------------------------------------------

export interface DocxDeleteArgs {
  doc_id: string;
  anchor?: string | undefined;
  range?: string | undefined;
  track_changes?: boolean | undefined;
  author?: string | undefined;
  response_format?: ResponseFormat | undefined;
}

export interface DocxDeleteResult {
  ok: boolean;
  deleted: number;
}

export function docxDelete(session: Session, args: DocxDeleteArgs): DocxDeleteResult {
  const doc = session.get(args.doc_id);
  if ((args.anchor == null) === (args.range == null)) {
    throw anchorInvalidError("Provide exactly one of anchor or range.");
  }
  const entries = paragraphEntries(doc);
  let targets: AnchorEntry[];
  if (args.anchor != null) {
    targets = [requireParagraph(entries, args.anchor)]; // hash validated FIRST
  } else {
    const range = args.range as string;
    const m = RANGE_RE.exec(range);
    if (!m) throw anchorInvalidError(`Malformed range string: ${range}.`);
    const start = Number(m[1]);
    const end = Number(m[3]);
    if (start > end) throw anchorInvalidError(`Inverted range: ${range}.`);
    const endpoints: [number, string | undefined][] = [
      [start, m[2]],
      [end, m[4]],
    ];
    for (const [ordinal, hash] of endpoints) {
      const entry = entryAt(entries, ordinal, `P${ordinal}`);
      if (hash !== undefined && entry.anchor !== `P${ordinal}#${hash}`) {
        throw anchorStaleError(`P${ordinal}#${hash}`);
      }
    }
    // The range deletes paragraphs a..b only; body tables between them are
    // untouched (table ops are Phase 2).
    targets = entries.slice(start - 1, end);
  }
  const xml = doc.documentXml();
  let newXml: string;
  if (args.track_changes !== true) {
    newXml = spliceAll(
      xml,
      targets.map((t) => ({ start: t.block.start, end: t.block.end, text: "" })),
    );
  } else {
    // §6a: one w:del per non-empty paragraph wrapping the full run content
    // after w:pPr (w:t renamed to w:delText; the paragraph mark survives).
    const author = revisionAuthor(args.author);
    const date = revisionDate();
    let revId = maxRevisionId(xml) + 1;
    const edits: SpliceEdit[] = [];
    for (const t of targets) {
      const p = t.block;
      if (p.selfClosed) continue; // counted, but nothing to wrap
      const pPr = childElements(xml, p.contentStart, p.contentEnd).find((k) => k.name === "w:pPr");
      const contentStart = pPr !== undefined ? pPr.end : p.contentStart;
      if (contentStart >= p.contentEnd) continue;
      edits.push({
        start: contentStart,
        end: p.contentEnd,
        text:
          revisionOpen("del", revId++, author, date) +
          tToDelText(xml.slice(contentStart, p.contentEnd)) +
          "</w:del>",
      });
    }
    newXml = spliceAll(xml, edits);
  }
  commit(doc, newXml);
  return { ok: true, deleted: targets.length };
}

// ---------------------------------------------------------------------------
// docx_revision (§7, §6a)
// ---------------------------------------------------------------------------

const REVISION_OPS = new Set(["list", "accept", "reject", "accept_all", "reject_all"]);

export interface DocxRevisionFilter extends RevisionFilter {}

export interface DocxRevisionArgs {
  doc_id: string;
  op: "list" | "accept" | "reject" | "accept_all" | "reject_all";
  id?: string | undefined;
  filter?: DocxRevisionFilter | undefined;
  response_format?: ResponseFormat | undefined;
}

export interface RevisionEntry {
  id: string;
  type: "ins" | "del";
  author: string;
  date: string;
  anchor?: string;
  text: string;
}

export interface DocxRevisionResult {
  revisions?: RevisionEntry[];
  accepted?: number;
  rejected?: number;
  remaining_by_author?: Record<string, number>;
  anchors?: string[];
}

export function docxRevision(session: Session, args: DocxRevisionArgs): DocxRevisionResult {
  const doc = session.get(args.doc_id);
  if (!REVISION_OPS.has(args.op)) {
    throw new ToolError("not_found", `Unknown revision op: ${String(args.op)}.`, [
      "Use list, accept, reject, accept_all, or reject_all.",
    ]);
  }
  const blocks = doc.anchorIndex();
  let xml = doc.documentXml();
  const revisions = scanRevisions(xml, blocks);
  const flt: RevisionFilter = args.filter ?? {};

  if (args.op === "list") {
    const listed: RevisionEntry[] = [];
    for (const rev of revisions) {
      if (!revisionMatches(rev, flt)) continue;
      listed.push({
        id: rev.id,
        type: rev.kind,
        author: rev.author,
        date: rev.date,
        ...(rev.anchor !== null ? { anchor: rev.anchor } : {}),
        text: rev.text,
      });
    }
    return { revisions: listed };
  }

  const accept = args.op === "accept" || args.op === "accept_all";
  let selected: Revision[];
  if (args.op === "accept_all" || args.op === "reject_all") {
    selected = [...revisions]; // §6a: _all ops ignore id/filter
  } else if (args.id != null) {
    // An id selecting nothing resolves nothing (§7 idempotency), not an error.
    selected = revisions.filter((rev) => rev.id === args.id);
  } else {
    selected = revisions.filter((rev) => revisionMatches(rev, flt));
  }
  // A candidate nested inside another candidate resolves with its container (§6a).
  const candidates = selected.filter(
    (rev) => !selected.some((other) => other.el.start < rev.el.start && rev.el.end < other.el.end),
  );
  const ordinals = [
    ...new Set(candidates.map((rev) => rev.ordinal).filter((o): o is number => o !== null)),
  ].sort((a, b) => a - b);
  if (candidates.length > 0) {
    xml = resolveRevisions(xml, candidates, accept);
    for (const ordinal of ordinals) xml = mergeParagraphRuns(xml, ordinal); // §7 post-pass
    commit(doc, xml);
  }
  const fresh = paragraphEntries(doc);
  const remaining = scanRevisions(doc.documentXml(), doc.anchorIndex());
  const byAuthor = new Map<string, number>();
  for (const rev of remaining) byAuthor.set(rev.author, (byAuthor.get(rev.author) ?? 0) + 1);
  const remainingByAuthor: Record<string, number> = {};
  for (const [author, n] of [...byAuthor].sort((a, b) =>
    a[0] < b[0] ? -1 : a[0] > b[0] ? 1 : 0,
  )) {
    remainingByAuthor[author] = n;
  }
  const anchors = ordinals.map((o) => entryAt(fresh, o, `P${o}`).anchor);
  return accept
    ? { accepted: candidates.length, remaining_by_author: remainingByAuthor, anchors }
    : { rejected: candidates.length, remaining_by_author: remainingByAuthor, anchors };
}

File: python/src/docxengine/_tools_edit.py
"""Edit-surface tools: docx_replace, docx_edit_paragraph, docx_insert, docx_delete,
docx_revision (algorithms.md §4–§7, §6a).

Each function takes the :class:`~docxengine._session.Session` plus keyword arguments
named exactly as in ``spec/tools/<tool>.json`` and returns the result object in that
schema's shape. Every edit validates its anchor hash first (``anchor_stale``) and
returns fresh anchors for everything it touched. Failures raise
:class:`~docxengine._errors.ToolError` with a ``spec/errors.json`` code.
"""

from __future__ import annotations

from . import _create, _edits, _xml
from ._anchors import build_anchor_index
from ._errors import ToolError
from ._session import Session

_REVISION_OPS = frozenset({"list", "accept", "reject", "accept_all", "reject_all"})

#: The §22 horizontal-rule paragraph (mirrors ``_create._emit_rule``).
_RULE_PARAGRAPH = (
    "<w:p><w:pPr><w:pBdr>"
    '<w:bottom w:val="single" w:sz="6" w:space="1" w:color="auto"/>'
    "</w:pBdr></w:pPr></w:p>"
)


def _plural(n: int, noun: str) -> str:
    return f"{n} {noun}" if n == 1 else f"{n} {noun}s"


def _parse_insert_blocks(content: str) -> list[_create.LineBlock]:
    """§6a insert markdown: classify each non-blank line (shares the §22 grammar).

    Headings, quotes, rules, task lists, ``-``/``*``/``1.`` list items, else plain;
    multi-line tables are not recognized here. Blank/whitespace-only lines emit nothing.
    """
    blocks: list[_create.LineBlock] = []
    for raw_line in content.split("\n"):
        line = raw_line.removesuffix("\r")
        if not line.strip(_xml.WHITESPACE):
            continue
        blocks.append(_create.classify_line(line))
    return blocks


def _block_style(block: _create.LineBlock) -> str | None:
    """The pStyle a §6a insert block carries (``None`` = no pStyle; ``numPr`` is Phase 2)."""
    if block.kind == "heading":
        return f"Heading{block.level}"
    if block.kind == "quote":
        return "Quote"
    if block.kind in ("task", "ul", "ol"):
        return "ListParagraph"
    return None


# ---------------------------------------------------------------------------
# docx_replace (§4/§5, §6a)
# ---------------------------------------------------------------------------


def docx_replace(
    session: Session,
    *,
    doc_id: str,
    old: str,
    new: str,
    anchor: str | None = None,
    all: bool = False,  # noqa: A002 - wire name pinned by the tool schema
    track_changes: bool = False,
    author: str | None = None,
    response_format: str = "concise",
) -> dict[str, object]:
    """Replace text in one anchored paragraph or across the whole body."""
    doc = session.get(doc_id)
    package = doc.package
    part = package.main_document_part()
    entries = _edits.paragraph_entries(package)
    target = _edits.require_paragraph(entries, anchor) if anchor is not None else None
    if not old:
        raise ToolError(
            "not_found",
            "Text not found: the search text is empty.",
            ["Provide non-empty old text."],
        )
    data = package.part(part)
    ordinals = [target.ordinal] if target is not None else [e.ordinal for e in entries]
    total = 0
    for ordinal in ordinals:
        raw, _ = _xml.paragraph_text(data, entries[ordinal - 1].span)
        total += _edits.count_occurrences(raw, old)
    if not all:
        if total == 0:
            raise ToolError(
                "not_found",
                f"Text not found: {old}.",
                ["Broaden the query; check the projection for the exact text."],
            )
        if total > 1:
            raise ToolError(
                "ambiguous_target",
                f"{old} matches {total} times without all: true.",
                ["Add all: true or narrow with an anchor."],
            )
    author_name = _edits.resolve_author(author)
    date = _edits.revision_date()
    n_replaced = 0
    affected: list[int] = []
    for ordinal in ordinals:
        pos = 0
        while True:
            data = package.part(part)
            span = _edits.paragraph_span_at(data, ordinal)
            raw, pieces = _xml.paragraph_text(data, span)
            start = raw.find(old, pos)
            if start < 0:
                break
            end = start + len(old)
            if track_changes:
                new_data = _edits.splice_replace_tracked(
                    data, pieces, start, end, new, author_name, date
                )
            else:
                new_data = _edits.splice_replace_plain(data, pieces, start, end, new)
            package.set_part(part, new_data)
            n_replaced += 1
            pos = start + len(new)
            if ordinal not in affected:
                affected.append(ordinal)
            if not all:
                break
        if not all and n_replaced:
            break
    if n_replaced:
        doc.mark_dirty()
    fresh = _edits.paragraph_entries(package)
    if all:  # §6a: all → anchors of affected paragraphs ascending; otherwise new_anchor
        return {
            "n_replaced": n_replaced,
            "anchors": [fresh[ordinal - 1].anchor for ordinal in sorted(affected)],
        }
    ordinal = target.ordinal if target is not None else affected[0]
    return {"n_replaced": n_replaced, "new_anchor": fresh[ordinal - 1].anchor}


# ---------------------------------------------------------------------------
# docx_edit_paragraph (§6, §6a)
# ---------------------------------------------------------------------------


def docx_edit_paragraph(
    session: Session,
    *,
    doc_id: str,
    anchor: str,
    text: str,
    track_changes: bool = False,
    author: str | None = None,
    response_format: str = "concise",
) -> dict[str, object]:
    """Rewrite a paragraph's full text; tracking emits the §6 minimal word-level redline."""
    doc = session.get(doc_id)
    package = doc.package
    part = package.main_document_part()
    entries = _edits.paragraph_entries(package)
    entry = _edits.require_paragraph(entries, anchor)
    data = package.part(part)
    raw, _ = _xml.paragraph_text(data, entry.span)
    ops = _edits.word_diff(_edits.diff_units(raw), _edits.diff_units(text))
    blocks = _edits.diff_blocks(ops)
    new_data = _edits.rebuild_paragraph(
        data,
        entry.span,
        blocks,
        tracked=track_changes,
        author=_edits.resolve_author(author),
        date=_edits.revision_date(),
    )
    package.set_part(part, new_data)
    doc.mark_dirty()
    fresh = _edits.paragraph_entries(package)
    changed = max(
        sum(1 for op, _ in ops if op == "del"),
        sum(1 for op, _ in ops if op == "ins"),
    )
    return {
        "new_anchor": fresh[entry.ordinal - 1].anchor,
        "diff": f"~{_plural(changed, 'word')} changed",
    }


# ---------------------------------------------------------------------------
# docx_insert (§6a)
# ---------------------------------------------------------------------------


def docx_insert(
    session: Session,
    *,
    doc_id: str,
    content: str,
    after: str | None = None,
    before: str | None = None,
    style: str | None = None,
    track_changes: bool = False,
    author: str | None = None,
    response_format: str = "concise",
) -> dict[str, object]:
    """Insert new paragraphs (plain text or minimal Markdown) after/before an anchor."""
    doc = session.get(doc_id)
    package = doc.package
    part = package.main_document_part()
    if (after is None) == (before is None):
        raise _edits.anchor_invalid_error("Provide exactly one of after or before.")
    entries = _edits.paragraph_entries(package)
    entry = _edits.require_paragraph(entries, after if after is not None else before or "")
    blocks = _parse_insert_blocks(content)
    style_id = _edits.resolve_style_id(package, style) if style is not None else None
    if not blocks:
        return {"new_anchors": []}
    data = package.part(part)
    author_name = _edits.resolve_author(author)
    date = _edits.revision_date()
    rev_id = _edits.next_revision_id(data) if track_changes else 0
    pieces: list[str] = []
    for block in blocks:
        # A horizontal rule has no runs; a style override makes it a styled empty paragraph.
        if block.kind == "rule" and style_id is None:
            pieces.append(_RULE_PARAGRAPH)
            continue
        effective_style = style_id if style_id is not None else _block_style(block)
        ppr = (
            f'<w:pPr><w:pStyle w:val="{_xml.escape_attr(effective_style)}"/></w:pPr>'
            if effective_style is not None
            else ""
        )
        text = "" if block.kind == "rule" else block.text
        runs = _create.emit_inline(text) if text else ""
        if track_changes and runs:
            runs = _edits.revision_open("ins", rev_id, author_name, date) + runs + "</w:ins>"
            rev_id += 1
        pieces.append(f"<w:p>{ppr}{runs}</w:p>")
    position = entry.span.end if after is not None else entry.span.start
    new_data = _xml.splice(data, [(position, position, "".join(pieces).encode("utf-8"))])
    package.set_part(part, new_data)
    doc.mark_dirty()
    fresh = _edits.paragraph_entries(package)
    base = entry.ordinal + 1 if after is not None else entry.ordinal
    return {"new_anchors": [fresh[base - 1 + i].anchor for i in range(len(blocks))]}


# ---------------------------------------------------------------------------
# docx_delete (§6a)
# ---------------------------------------------------------------------------


def docx_delete(
    session: Session,
    *,
    doc_id: str,
    anchor: str | None = None,
    range: str | None = None,  # noqa: A002 - wire name pinned by the tool schema
    track_changes: bool = False,
    author: str | None = None,
    response_format: str = "concise",
) -> dict[str, object]:
    """Delete one paragraph or a contiguous range; tracked deletion is a redline."""
    doc = session.get(doc_id)
    package = doc.package
    part = package.main_document_part()
    if (anchor is None) == (range is None):
        raise _edits.anchor_invalid_error("Provide exactly one of anchor or range.")
    entries = _edits.paragraph_entries(package)
    if anchor is not None:
        targets = [_edits.require_paragraph(entries, anchor)]
    else:
        assert range is not None
        m = _edits.RANGE_RE.match(range)
        if not m:
            raise _edits.anchor_invalid_error(f"Malformed range string: {range}.")
        start, end = int(m.group(1)), int(m.group(3))
        if start > end:
            raise _edits.anchor_invalid_error(f"Inverted range: {range}.")
        for ordinal, hash_part in ((start, m.group(2)), (end, m.group(4))):
            entry = _edits.entry_at(entries, ordinal, f"P{ordinal}")
            if hash_part is not None and entry.anchor != f"P{ordinal}#{hash_part}":
                raise _edits.anchor_stale_error(f"P{ordinal}#{hash_part}")
        targets = entries[start - 1 : end]
    data = package.part(part)
    if not track_changes:
        new_data = _xml.splice(data, [(t.span.start, t.span.end, b"") for t in targets])
    else:
        author_name = _edits.resolve_author(author)
        date = _edits.revision_date()
        rev_id = _edits.next_revision_id(data)
        edits: list[tuple[int, int, bytes]] = []
        for target in targets:
            p = target.span
            if p.empty:
                continue
            ppr = next(
                _xml.iter_elements(data, p.inner_start, p.inner_end, names=("w:pPr",), max_depth=1),
                None,
            )
            content_start = ppr.end if ppr is not None else p.inner_start
            if content_start >= p.inner_end:
                continue
            wrapped = (
                _edits.revision_open("del", rev_id, author_name, date).encode("utf-8")
                + _edits.t_to_deltext(data[content_start : p.inner_end])
                + b"</w:del>"
            )
            edits.append((content_start, p.inner_end, wrapped))
            rev_id += 1
        new_data = _xml.splice(data, edits)
    package.set_part(part, new_data)
    doc.mark_dirty()
    return {"ok": True, "deleted": len(targets)}


# ---------------------------------------------------------------------------
# docx_revision (§7, §6a)
# ---------------------------------------------------------------------------


def docx_revision(
    session: Session,
    *,
    doc_id: str,
    op: str,
    id: str | None = None,  # noqa: A002 - wire name pinned by the tool schema
    filter: dict[str, str] | None = None,  # noqa: A002 - wire name pinned by the tool schema
    response_format: str = "concise",
) -> dict[str, object]:
    """List, accept, or reject tracked changes, individually, in bulk, or filtered."""
    doc = session.get(doc_id)
    package = doc.package
    part = package.main_document_part()
    if op not in _REVISION_OPS:
        raise ToolError(
            "not_found",
            f"Unknown revision op: {op}.",
            ["Use list, accept, reject, accept_all, or reject_all."],
        )
    blocks = build_anchor_index(package)
    data = package.part(part)
    revisions = _edits.scan_revisions(data, blocks)
    flt = dict(filter or {})
    if op == "list":
        listed: list[dict[str, str]] = []
        for rev in revisions:
            if not _edits.revision_matches(rev, flt):
                continue
            item = {"id": rev.rev_id, "type": rev.kind, "author": rev.author, "date": rev.date}
            if rev.anchor is not None:
                item["anchor"] = rev.anchor
            item["text"] = rev.text
            listed.append(item)
        return {"revisions": listed}
    accept = op in ("accept", "accept_all")
    if op in ("accept_all", "reject_all"):  # §6a: _all ops ignore id/filter
        candidates = list(revisions)
    elif id is not None:
        # An id selecting nothing resolves nothing (§7 idempotency), not an error.
        candidates = [rev for rev in revisions if rev.rev_id == id]
    else:
        candidates = [rev for rev in revisions if _edits.revision_matches(rev, flt)]
    candidates = [  # a candidate nested in another candidate resolves with its container
        rev
        for rev in candidates
        if not any(
            other.span.start < rev.span.start and rev.span.end < other.span.end
            for other in candidates
        )
    ]
    ordinals = sorted({rev.ordinal for rev in candidates if rev.ordinal is not None})
    if candidates:
        new_data = _edits.resolve_revisions(data, candidates, accept)
        for ordinal in ordinals:
            new_data = _edits.merge_paragraph_runs(new_data, ordinal)
        package.set_part(part, new_data)
        doc.mark_dirty()
    fresh_blocks = build_anchor_index(package)
    fresh = [b for b in fresh_blocks if b.kind == "paragraph"]
    remaining = _edits.scan_revisions(package.part(part), fresh_blocks)
    by_author: dict[str, int] = {}
    for rev in remaining:
        by_author[rev.author] = by_author.get(rev.author, 0) + 1
    return {
        "accepted" if accept else "rejected": len(candidates),
        "remaining_by_author": dict(sorted(by_author.items())),
        "anchors": [fresh[ordinal - 1].anchor for ordinal in ordinals],
    }
